feat: better gps and gcp align

YanNoun opened the pull request:
This PR aims at improving the alignment behaviour when using GPS and GCP.

More specifically, we propagate the use of `bundle_use_gps` and `bundle_use_gcp` in the alignment. Discussion is welcome whether we should have a distinct option or not.

The global behaviour is also changed as it is done before any bundle, hoping GPS/GCP error will be closer to optimum when starting the bundle. Additionally, we also support edge cases when we have only one or two GCP/GPS (scale/orientation ambiguity will remain though, no a-priori is being made).

Finally, we refactored code a bit so we can support correct topocentric frame when there is no GPS.

Yann

YanNoun commented:
Let's add orientation prior in the alignment.

pierotofy commented on opensfm/align.py:
```diff
@@ -55,26 +55,38 @@ def align_reconstruction_similarity(reconstruction, gcp, config):
+
+    # For now, translation only, need to decide on a prior rotation
+    if len(X) == 2:
+        s = np.linalg.norm(Xp[0] - Xp[1])/np.linalg.norm(np.array(X[0]) - np.array(X[1]))
```
This should be:

`s = np.linalg.norm(np.array(Xp[0]) - np.array(Xp[1]))/np.linalg.norm(np.array(X[0]) - np.array(X[1]))`

Per the problem that was discussed in #413 (see comment from Jun 20).

pierotofy commented:
Finally had time to test this (sorry it took so long). @YanNoun 

Running this branch with these input images + GCP https://github.com/OpenDroneMap/odm_data_copr still produces a reconstruction that is out of scale:

![image](https://user-images.githubusercontent.com/1951843/62955309-ce0e7280-bdbe-11e9-8963-dfa9109abd59.png)

The green/red/blue dots are the cameras exported with `export_ply` (so the reconstruction looks really tiny). I'm going to investigate this a bit further to see if I can find a cause/fix.

pierotofy commented:
Using:

```
align_method: naive
```

Scale looks OK, but rotation/translation does not.

![image](https://user-images.githubusercontent.com/1951843/62976194-09726680-bdea-11e9-83ae-d90e827f2948.png)

![image](https://user-images.githubusercontent.com/1951843/62976242-2c9d1600-bdea-11e9-9d92-b744128fcf20.png)

pierotofy commented:
Interestingly, the GCP locations seem to be triangulated fine.

![image](https://user-images.githubusercontent.com/1951843/62976807-6de1f580-bdeb-11e9-99aa-29c5c22a508a.png)

I've extracted the GCP coordinates by adding a `print` in `_add_gcp_to_bundle`:

`reconstruction.py`
...
        coordinates = triangulate_gcp(point, shots)
        if coordinates is None:
            if point.coordinates is not None:
                print(point.coordinates)
                coordinates = point.coordinates
            else:
                logger.warning("Cannot initialize GCP '{}'."
                               "  Ignoring it".format(point.id))
                continue
...

```
[-4.85703576e+00  4.65463388e+00 -3.55299562e-06]
[ 1.44868850e+01 -1.80340258e+01 -4.20194119e-05]
[-9.43289669e-01 -5.33349274e+00 -2.30874866e-06]
[-1.61153324e+01  1.79837399e+01 -4.57800925e-05]
[ 6.54344838e+00 -1.05103216e+01 -1.20438635e-05]
[ 7.15283159e-01  4.91304974e+00 -1.93994492e-06]
[-1.77048913e+01  1.12889474e+01 -3.45725566e-05]
[ 6.40999243e+00 -5.46441798e+00 -5.56744635e-06]
[-6.28309068e+00  1.36635485e+01 -1.77789479e-05]
[ 1.77481583e+01 -1.31615936e+01 -3.82959843e-05]
2019-08-13 20:17:52,976 DEBUG: Ceres Solver Report: Iterations: 4, Initial cost: 1.261175e+02, Final cost: 1.245557e+02, Termination: CONVERGENCE
2019-08-13 20:17:53,140 INFO: Removed outliers: 1
2019-08-13 20:17:53,169 INFO: {'average_track_length': 2.3795516546079942, 'points_count': 8431, 'observations_count': 20062, 'cameras_count': 3, 'average_track_length_notwo': 3.0}
2019-08-13 20:17:53,169 INFO: Reconstruction 0: 38 images, 80350 points
2019-08-13 20:17:53,169 INFO: Reconstruction 1: 3 images, 8431 points
2019-08-13 20:17:53,169 INFO: 2 partial reconstructions in total.
```

pierotofy commented:
(off-topic comment, I've got BOW matching to work, it's AWESOME! )

pierotofy commented:
Can't seem to find anything too obvious. But seems like the last call to `bundle` fails to converge to a solution:

```
2019-08-14 02:50:09,600 DEBUG: Ceres Solver Report: Iterations: 51, Initial cost: 1.318168e+05, Final cost: 5.121470e+04, Termination: NO_CONVERGENCE
2019-08-14 02:50:15,185 INFO: Removed outliers: 785
2019-08-14 02:50:15,441 INFO: {'average_track_length': 3.75419755560008, 'points_count': 74865, 'observations_count': 281058, 'cameras_count': 38, 'average_track_length_notwo': 4.869867354297328}
```

Please let me know if I can provide additional information / test data. :pray:

YanNoun commented:
Hi @pierotofy ,

Spend some time investigating what's going on. First some observations :
 - First of all, there is only 2 GCPs having at least 2 projections, thus being triangulated (all the others GCP have one projection). We recommend at least 3 GCPs with >= 2 projections (let's call them triangulated GCPs) for proper alignment.
 - Thanks to the `orientation_prior`, OpenSfM can align even with 1 or 2 triangulated GCPs.

Then, what I fixed :
 - There was an obvious bug (not solving for the rotation) in the 2 triangulated GCPs case. It is solved now.

Then, the most problematic point, the data :
 - You can use the command `./bin/plot_gcp` to see the GCP projected (blue) in the reconstructed image versus its annotated position in the image (orange). In our case, the alignment of the two triangulated GCPs will be perfect (exact solution, orange and blue superposed) but you will see some huge discrepancies on the others GCPs (1 projections ones), though, which leads me to the next point.
 - I never managed to get any consistent alignment with 3 or 4 triangulated GCPs (I added some more projections). I processed the project with Agisoft/Metashape as reference with the same GCP data and I had error of 1 meter. Both with `WGS84` and `UTM` coordinates. See below (`source` mean GCP position, other being the estimated position). So I wonder what's wrong with the data itself (what's the accuracy of the surveyed data ?).
![agisoft_odm](https://user-images.githubusercontent.com/39743653/63185300-f8c82780-c059-11e9-908b-2491c4b5fc03.jpg)
 - I had the same magnitude of 3D error (1 meter) in OpenSfM and Agisoft. Somehow a good news.

I finally added a feature to not bother anymore with alignment method :
 - I added a `auto` mode as `align_method` which will automatically switch between `naive` and `orientation_prior` depending on the GCP configuration (if they lie on a line or not).

Would be good if we could test on some other dataset with better GCPs.

Yann

pierotofy commented:
Thanks for looking into this @YanNoun ! Helpful as always.

I think your observations are correct, I don't think these measurements were collected with precision equipment. @smathermather might know the background story for how this data was collected, but I think it's safe to assume an error of a few meters, which is within the results obtained. Like you mentioned, only two GCPs are tied to two images, which is not ideal.

With the latest changes and using `auto` (nice!) the result looks at scale and in the right place:

![image](https://user-images.githubusercontent.com/1951843/63196160-9ca5d900-c042-11e9-996e-c002afed4efb.png)

:tada: 

I will test two more datasets that I know have better accuracy and post my results here.

pierotofy commented:
Looks good on a dataset with both GPS and GCPs.

![image](https://user-images.githubusercontent.com/1951843/63206325-540d1080-c080-11e9-9cfd-4a956b2e1d8c.png)

pierotofy commented:
And here's another one:

![image](https://user-images.githubusercontent.com/1951843/63212852-2ce22d80-c0d8-11e9-9c61-10d60724d394.png)

![image](https://user-images.githubusercontent.com/1951843/63212859-39668600-c0d8-11e9-9eb9-7caad6b96707.png)

![image](https://user-images.githubusercontent.com/1951843/63212863-45eade80-c0d8-11e9-92f2-11e948ef126a.png)

![image](https://user-images.githubusercontent.com/1951843/63212865-50a57380-c0d8-11e9-8f27-7fd4ceff08b7.png)

![image](https://user-images.githubusercontent.com/1951843/63212866-5d29cc00-c0d8-11e9-8a51-0f050b96c012.png)

pierotofy commented:
Would be helpful, yes.

denistestemale commented:
Hello.
Sorry for commenting so late. One of the dataset was mine, obtained with <2cm precision. With a recent WebODM version (updated april 28th 2020) the results are very good indeed:
- precision is now below 8cm for all GCPs, on several datasets obtained with the same GPS;
- doming effect disappeared (might be related). 

Thanks!
d
